app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkFragmentInteractor.kt

```diff
@@ -0,0 +1,172 @@
+
+package org.mozilla.fenix.library.bookmarks
+
+import android.content.Context
+import android.view.View
+import androidx.navigation.NavController
+import mozilla.components.concept.storage.BookmarkNode
+import mozilla.components.concept.storage.BookmarkNodeType
+import org.mozilla.fenix.BrowserDirection
+import org.mozilla.fenix.BrowsingModeManager
+import org.mozilla.fenix.HomeActivity
+import org.mozilla.fenix.R
+import org.mozilla.fenix.components.FenixSnackbar
+import org.mozilla.fenix.components.metrics.Event
+import org.mozilla.fenix.components.metrics.MetricController
+import org.mozilla.fenix.ext.asActivity
+import org.mozilla.fenix.ext.components
+import org.mozilla.fenix.ext.copyUrl
+import org.mozilla.fenix.ext.nav
+
+@SuppressWarnings("TooManyFunctions")
+class BookmarkFragmentInteractor(
+    private val context: Context,
+    private val navController: NavController,
+    private val bookmarkStore: BookmarkStore,
+    private val sharedViewModel: BookmarksSharedViewModel,
+    private val deleteBookmarkNodes: (Set<BookmarkNode>, Event) -> Unit
+) : BookmarkInteractor, SignInInteractor {
+
+    val activity: HomeActivity?
+        get() = context.asActivity() as? HomeActivity
+    val metrics: MetricController
+        get() = context.components.analytics.metrics
+
+    override fun change(node: BookmarkNode) {
+        bookmarkStore.dispatch(BookmarkAction.Change(node))
+    }
+
+    override fun open(item: BookmarkNode) {
+        if (item.type == BookmarkNodeType.ITEM) {
+            item.url?.let { url ->
+                activity!!
+                    .openToBrowserAndLoad(
+                        searchTermOrURL = url,
+                        newTab = true,
+                        from = BrowserDirection.FromBookmarks
+                    )
+            }
+        }
+        metrics.track(Event.OpenedBookmark)
+    }
+
+    override fun expand(folder: BookmarkNode) {
+        navController.nav(
+            R.id.bookmarkFragment,
+            BookmarkFragmentDirections.actionBookmarkFragmentSelf(folder.guid)
+        )
+    }
+
+    override fun switchMode(mode: BookmarkState.Mode) {
+        activity?.invalidateOptionsMenu()
+    }
+
+    override fun edit(node: BookmarkNode) {
+        navController.nav(
+            R.id.bookmarkFragment,
+            BookmarkFragmentDirections
+                .actionBookmarkFragmentToBookmarkEditFragment(node.guid)
+        )
+    }
+
+    override fun select(node: BookmarkNode) {
+        bookmarkStore.dispatch(BookmarkAction.Select(node))
+    }
+
+    override fun deselect(node: BookmarkNode) {
+        bookmarkStore.dispatch(BookmarkAction.Deselect(node))
+    }
+
+    override fun deselectAll() {
+        clearBookmarkSelection()
+    }
+
+    override fun copy(item: BookmarkNode, view: View?) {
+        item.copyUrl(activity!!)
+        if (view != null) {
+            FenixSnackbar.make(view, FenixSnackbar.LENGTH_LONG).setText(context.getString(R.string.url_copied)).show()
```

Comment:
Creating a `FenixSnackBar` in the interactor doesn't look right. We could create a presenter (Even just a FenixSnackbarPresenter interface?) and pass that in to the interactor. That would make this easier to add to the tests as well.

Reply:
I'll create a simple presenter then.